contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,196 @@
+	// Must match in order to be controlled.
+	// If empty, defaulted to labels on Pod template.
+	// More info: http://kubernetes.io/docs/user-guide/labels#label-selectors
+	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,1,opt,name=selector"`
+
+	// Template is the object that describes the pod that will be created.
+	// The DaemonSet will create exactly one copy of this pod on every node
+	// that matches the template's node selector (or on every node if no node
+	// selector is specified).
+	// More info: http://kubernetes.io/docs/user-guide/replication-controller#pod-template
+	Template v1.PodTemplateSpec `json:"template" protobuf:"bytes,2,opt,name=template"`
+
+	// Update strategy to replace existing DaemonSet pods with new pods.
+	UpdateStrategy DaemonSetUpdateStrategy `json:"updateStrategy,omitempty"`
+}
+
+const (
+	// DefaultDaemonSetUniqueLabelKey is the default key of the labels that is added
+	// to daemon set pods to distinguish between old and new pod templates during
+	// DaemonSet update. See DaemonSetSpec's UniqueLabelKey field for more information.
+	DefaultDaemonSetUniqueLabelKey string = "daemonset.kubernetes.io/podTemplateHash"
+)
+
+// DaemonSetStatus represents the current status of a daemon set.
+type DaemonSetStatus struct {
+	// CurrentNumberScheduled is the number of nodes that are running at least 1
+	// daemon pod and are supposed to run the daemon pod.
+	CurrentNumberScheduled int32
+
+	// NumberMisscheduled is the number of nodes that are running the daemon pod, but are
+	// not supposed to run the daemon pod.
+	NumberMisscheduled int32
+
+	// DesiredNumberScheduled is the total number of nodes that should be running the daemon
+	// pod (including nodes correctly running the daemon pod).
+	DesiredNumberScheduled int32
+
+	// NumberReady is the number of nodes that should be running the daemon pod and have one
+	// or more of the daemon pod running and ready.
+	NumberReady int32
+
+	// ObservedGeneration is the most recent generation observed by the daemon set controller.
+	ObservedGeneration int64
+
+	// UpdatedNumberScheduled is the total number of nodes that are running updated
+	// daemon pod
+	UpdatedNumberScheduled int32 `json:"updatedNumberScheduled"`
+
+}
+```
+
+### Controller 
+
+#### DaemonSet Controller 
+
+The DaemonSet Controller will make DaemonSet updates happen. It will watch
+DaemonSets in etcd. 
+
+For each pending DaemonSet updates, it will:
+
+1. Find all pods whose label is matched by `DaemonSetSpec.Selector`. 
+1. Find all nodes that should run these pods created by this DaemonSet.
+1. Find an existing PodTemplate whose `Template` is the same as
+   `DaemonSetSpec.Template`
+   - If not found, create a new PodTemplate from `DaemonSetSpec.Template`
+   - The name will be `<DaemonSet-Name>-<Hash-of-pod-template>`
+   - Add "pod-template-hash" labels to `PodTemplate.Metadata.Labels`, value be
+     the hash of `PodTemplate.Spec`
```

Comment:
When we talked about this, I think it has to be a separate object to store the history because we cannot store more than 1 copy of spec inside DS without starting to run out of size.  If we don't think PodTemplate is right, the alternative is DaemonSetRevision, which sets a precedent for StatefulSetRevision (remember, stateful sets don't use *just* PodTemplate).